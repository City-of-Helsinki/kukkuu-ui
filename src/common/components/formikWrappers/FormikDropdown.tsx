import React from 'react';
import { Select, SelectProps as HDSDropdownProps } from 'hds-react';
import { useField } from 'formik';
import { useTranslation } from 'react-i18next';

import { Option } from './types';
import styles from './formikInputs.module.scss';

type DropdownProps = Omit<
  HDSDropdownProps<Option>,
  'value' | 'onChange' | 'options' | 'defaultValue'
>;

type Props = DropdownProps & {
  name: string;
  onChange?: (selected: Option) => void;
  value?: string;
  options: Option[];
};

const emptyValue = {
  label: '',
  value: '',
};

function FormikDropdown({
  name,
  onChange = () => ({}),
  value: userValue,
  options,
  ...rest
}: Props) {
  const { t } = useTranslation();
  const [{ value, ...field }, meta, helpers] = useField(name);

  const handleChange = (selectedItem: Option | Option[]) => {
    if (Array.isArray(selectedItem)) {
      helpers.setValue(selectedItem.map((item) => item.value));
    } else {
      helpers.setValue(selectedItem.value);
    }
  };

  const handleBlur = () => {
    helpers.setTouched(true);
  };

  const usedValue = userValue || value;
  const valueAsOption = options.find((option) => option.value === usedValue);

  return (
    <Select<Option>
      {...field}
      className={styles.formField}
      value={valueAsOption || emptyValue}
      options={options}
      onChange={handleChange}
      invalid={meta.touched && Boolean(meta.error)}
      error={meta.touched && Boolean(meta.error) && t(meta.error || '')}
      onBlur={handleBlur}
      {...rest}
      multiselect={false}
    />
  );
}

export default FormikDropdown;
