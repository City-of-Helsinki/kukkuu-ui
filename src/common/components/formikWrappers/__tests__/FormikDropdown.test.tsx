import React from 'react';
import { shallow } from 'enzyme';

import FormikDropdown from '../FormikDropdown';
import TestForm from '../../../test/TestForm';

it('renders snapshot correctly', () => {
  const props = {
    id: 'fooSelect',
    name: 'fooSelect',
    label: 'select label',
    onChange: () => jest.fn(),
    options: [
      {
        label: 'foo_label',

        value: 'foo',
      },
      {
        label: 'bar_label',
        value: 'bar',
      },
    ],
  };
  const input = shallow(
    <TestForm>{() => <FormikDropdown {...props} />}</TestForm>
  );
  expect(input.html()).toMatchSnapshot();
});
