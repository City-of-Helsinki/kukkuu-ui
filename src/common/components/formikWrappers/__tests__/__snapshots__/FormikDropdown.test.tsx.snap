// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders snapshot correctly 1`] = `"<div class=\\"Dropdown-module_root__2LI4-\\"><label for=\\"fooSelect-toggle-button\\" class=\\"FieldLabel-module_label__2tl-4 text-input_hds-text-input__label__pDw_W \\" id=\\"fooSelect-label\\">select label</label><div class=\\"Dropdown-module_wrapper__1oIw5\\"><button type=\\"button\\" id=\\"fooSelect-toggle-button\\" aria-haspopup=\\"listbox\\" aria-expanded=\\"false\\" aria-labelledby=\\"fooSelect-label fooSelect-toggle-button\\" class=\\"Dropdown-module_buttonDropdown__IwbnQ Dropdown-module_dropdown__22ZdF Dropdown-module_placeholder__2nbDo Dropdown-module_buttonReset__r__8I helpers_button-reset__QA3d0\\"><span class=\\"Dropdown-module_buttonLabel__2jVRU\\"></span><svg class=\\"Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik Dropdown-module_angleIcon__1Y1mT\\" viewBox=\\"0 0 24 24\\" xmlns=\\"http://www.w3.org/2000/svg\\"><g fill=\\"none\\" fill-rule=\\"evenodd\\"><path d=\\"M0 0h24v24H0z\\"></path><path fill=\\"currentColor\\" d=\\"M12 13.5l5-5 1.5 1.5-6.5 6.5L5.5 10 7 8.5z\\"></path></g></svg></button></div><ul id=\\"fooSelect-menu\\" role=\\"listbox\\" aria-labelledby=\\"fooSelect-label\\" tabindex=\\"-1\\" class=\\"Dropdown-module_menu__3rei6\\" style=\\"max-height:calc(var(--dropdown-height) * 5)\\"></ul></div>"`;
